live-data-collector/src/main/scala/pl/touk/nussknacker/engine/livedata/LiveDataUploaderHolder.scala

```diff
@@ -0,0 +1,86 @@
+package pl.touk.nussknacker.engine.livedata
+
+import com.github.benmanes.caffeine.cache.Caffeine
+import org.slf4j.LoggerFactory
+import pl.touk.nussknacker.engine.api.process.ProcessIdWithName
+import pl.touk.nussknacker.engine.livedata.LiveDataUploader.LiveDataUploaderConfig
+import pl.touk.nussknacker.engine.newdeployment.DeploymentId
+
+import java.time.Instant
+import java.util.concurrent.{Executors, ScheduledExecutorService, ScheduledFuture, TimeUnit}
+import scala.compat.java8.FunctionConverters.asJavaFunction
+
+private[livedata] object LiveDataUploaderHolder {
+
+  private val activeUploaders =
+    Caffeine
+      .newBuilder()
+      .expireAfterAccess(java.time.Duration.ofHours(1))
+      .build[
+        CacheKey,
+        String
+      ]() // Should be [(ProcessIdWithName, DeploymentId), Unit], but Scala 2.12 does not accept tuples and Unit
+
+  private val logger = LoggerFactory.getLogger(getClass)
+
+  def getExistingOrStartLiveDataUploader(
+      processIdWithName: ProcessIdWithName,
+      deploymentIdOpt: Option[DeploymentId],
+      config: LiveDataUploaderConfig,
+  ): Unit = {
+    deploymentIdOpt match {
+      case Some(deploymentId) =>
+        activeUploaders.get(
+          CacheKey(processIdWithName, deploymentId),
+          asJavaFunction((cacheKey: CacheKey) => {
+            startPeriodicLiveDataUploader(cacheKey.processIdWithName, cacheKey.deploymentId, config)
+            ""
+          })
+        )
+      case None =>
+        logger.error(
+          "Live data uploader cannot be started, because the DeploymentId is not defined or has non-UUID underlying value"
+        )
+    }
+  }
+
+  def startPeriodicLiveDataUploader(
+      processIdWithName: ProcessIdWithName,
+      deploymentId: DeploymentId,
+      config: LiveDataUploaderConfig,
+  ): Unit = {
+    logger.info(s"Starting live data uploader for scenario $processIdWithName with interval ${config.intervalSeconds}s")
+    val uploader                                  = new LiveDataUploader(config)
+    val scheduler: ScheduledExecutorService       = Executors.newSingleThreadScheduledExecutor()
+    var scheduledTask: Option[ScheduledFuture[_]] = None
+
+    def uploadLiveData(): Unit = {
+      LiveDataCollectingListenerHolder
+        .storageOpt(processIdWithName.name)
+        .foreach { storage =>
+          val shouldStop =
+            storage.getLastUpdatedAt < Instant.now.getEpochSecond - config.uploaderInactivityTimeoutInSeconds
+          if (shouldStop) {
+            logger.info(s"Stopping live data uploader for scenario $processIdWithName because of inactivity")
+            scheduledTask.foreach(_.cancel(true))
+            scheduler.shutdownNow()
+            activeUploaders.invalidate(CacheKey(processIdWithName, deploymentId))
+            uploader.close()
```

Comment:
We can create one uploader and one scheduler and reuse it between different deployments on the same task manager. With the current solution, you create many db connections, many threa pools and many threads, we can avoid that.

Reply:
can we do it (sharing uploader among jobs on the same taskmanager) without placing the uploader code in flink `lib` dir (to use AppClassloader)?

Reply:
It seems that classloaders are isolated between jobs. On standalone Flink the cache will only contain single active uploader.

In the team that handles the production deployment we started to brainstorm what needs to be done in order to run it for 100s of scenarios. In this basic solution for cloud/demo I think it may stay this way (separate uploader per scenario) and be later improved when clients need it.

Reply:
We discussed it with @raphaelsolarski offline and this static context is not shared between jobs, because the same classes created by different jobs have different memory area.
Based on this, in order to reuse db connection and threads, we'd need to extract a separate lib and add it to flink libs. I think that for now we can ignore this problem, and leave a TODO as a potential direction of changes.
It also implies that we don't need to have a Map scenarioId -> something because we can be sure that this holder is used only by one job, for one scenario. WDYTAT @mgoworko ?

Reply:
I added todo comment. 

I this we should still have some kind of map (at the moment it is caffeine cache). My argument is, that from the purely JVM point of view, this code should work for multiple scenarios at the same time. It's the Flink design details, that prevent it from working that way. But this project module is not only Flink-specific. It might be reused for other DM's, for example RR (we might need to do that soon if client requests it). Changing the type from caffeine cache (or any kind of map) to single value represented by Option/Atomic/sth similar, will make it not work correctly in all other non-Flink contexts.